include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1139,11 +1168,17 @@ __parallel_find_or(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPoli
     assert(__rng_n > 0);
 
     // TODO: find a way to generalize getting of reliable work-group size
-    auto __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+    std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
 #if _ONEDPL_COMPILE_KERNEL
     auto __kernel = __internal::__kernel_compiler<_FindOrKernel>::__compile(__exec);
     __wgroup_size = ::std::min(__wgroup_size, oneapi::dpl::__internal::__kernel_work_group_size(__exec, __kernel));
 #endif
+
+#if _ONEDPL_FPGA_DEVICE
+    // Limit the maximum work-group size to minimize the cost of work-group reduction.
+    // Limiting this also helps to avoid huge work-group sizes on some devices (e.g., FPGU emulation).
+    __wgroup_size = std::min(__wgroup_size, (std::size_t)2048);
+#endif
```

Comment:
This limit was not intended for FPGA only. We don't want overly large work-groups on any device to not slow down the work-group reduction too much and prevent running in potential SLM limits. I would propose removing the if-statement. We might even limit this to 1024, which is picked on PVC anyways.

Reply:
In this PR I hadn't the goal to define `__wgroup_size` by any other way. It's changed here under this define (initially in your PR https://github.com/oneapi-src/oneDPL/pull/1667, thanks @julianmi) just for avoid long compilation time.
Let's do any changes about `__wgroup_size` in separate PRs.

Reply:
The issue with the test timeouts is not specific to FPGAs. This issue occurs on all device that support very large work-groups. This leads to more complex reductions which typically require more SLM. This slows the execution down and might even lead to timeouts as we've experienced. So, I don't think this is a compilation issue.

The proposed change in https://github.com/oneapi-src/oneDPL/pull/1667 limits the group size to a reasonable limit and should be safe on all architectures. Typical CPUs and GPUs group sizes are below this threshold.
 
 What other changes to the work-group size would you propose in another PR?

Reply:
I propose to remove in the separate PR `_ONEDPL_FPGA_DEVICE` define here and change the maximal available value from `2048` to something else.

Reply:
Yes, I agree that the change should be for all architectures as was done in #1667.  Generally, I think this is a good idea for most all of our patterns, as they are written with this sort of size in mind for the SYCL backend and it can lead to unexpected results for CPU and FPGA targets. Of course, there may be some reason for a specific algorithm to want larger workgroups when they are available.
  
I wonder if we would benefit from adding some cap like this within `__max_work_group_size` with some defaulted parameter which could be adjusted when specifically desired.  I think this could end up stopping these unexpected slowdowns / issues on non-GPU SYCL targets.

Reply:
> I wonder if we would benefit from adding some cap like this within `__max_work_group_size` with some defaulted parameter which could be adjusted when specifically desired. I think this could end up stopping these unexpected slowdowns / issues on non-GPU SYCL targets.

I agree with this. One drawback I see with adding this to `__max_work_group_size` is that it would require significant testing to not introduce any unwanted side effects. How about we introduce a helper such as `__limit_work_group_size` first for places where we explicitly limit this (e.g., `__parallel_find_or`, `__parallel_histogram_select_kernel`, `__parallel_transform_reduce`)? We could extend this over time and finally add something to `__max_work_group_size` or a similar generic helper.

Reply:
Yes, something like that may be the better approach, and allow us to more deliberately add this to patterns while looking at them specifically, rather than just making a blanket change "silently".